lib/tree.rb

```diff
@@ -17,41 +17,139 @@ def initialize
   end
 
   # Time Complexity: 
-  # Space Complexity: 
+  # Space Complexity: o(n)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else 
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end
+
+
+#   def add(key, value)
+#   if @root.nil?
+#     @root = TreeNode.new(key, value)
+#   else
+#     current = @root
+#     while true
+#       if key < current.key
+#         if !current.left.nil?
+#           current = current.left
+#         else
+#           current.left = TreeNode.new(key, value)
+#         end
+#       else
+#         if !current.right.nil?
+#           current = current.right
+#         else
+#           current.right = TreeNode.new(key, value)
+#         end
+#       end
+#     end
+#   end
+# end
+
+
   # Time Complexity: 
-  # Space Complexity: 
+  # Space Complexity: o(n)
   def find(key)
```

Comment:
The space/time complexity is O(log n) if the tree is balanced and O(n) if it's not.